sylt-compiler/src/typechecker.rs

```diff
@@ -1362,14 +1360,19 @@ impl TypeChecker {
         self.add_constraint(defined_ty, span, Constraint::Variable);
 
         match &value.kind {
-            ExpressionKind::Function { params, .. } => {
+            ExpressionKind::Function { params, pure, .. } => {
                 // Special case for functions
                 let args = params
                     .iter()
                     .map(|_| self.push_type(Type::Unknown))
                     .collect();
                 let ret = self.push_type(Type::Unknown);
-                let fn_ty = self.push_type(Type::Function(args, ret));
+                let purity = if *pure {
+                    Purity::Pure
+                } else {
+                    Purity::Undefined
+                };
```

Comment:
you know, with a `struct Pure(bool)` this would become

```
let purity = pure.then(|| Pure(true));
```

Reply:
I kind of did this, but without the struct. To me `struct Pure(bool)` seems kind of weird.